crates/vm/wasm/src/environment.rs

```diff
@@ -273,3 +273,415 @@ impl Environment {
+
+#[cfg(test)]
+mod test {
+    use {
+        crate::{Environment, Iterator, VmError, VmResult, WasmVm, GAS_PER_OPERATION},
+        grug_app::{GasTracker, QuerierProvider, Shared, StorageProvider},
+        grug_types::{
+            BlockInfo, Hash256, MockStorage, NumberConst, Order, StdError, Storage, Timestamp,
+            Uint64,
+        },
+        std::sync::Arc,
+        test_case::test_case,
+        wasmer::{
+            imports, CompilerConfig, Engine, Instance, Module, RuntimeError, Singlepass, Store,
+            Value,
+        },
+        wasmer_middlewares::{metering::set_remaining_points, Metering},
+    };
+
+    const MOCK_WAT: &[u8] = br#"(module (memory (export "memory") 1))"#;
+
+    const MOCK_BLOCK: BlockInfo = BlockInfo {
+        height: Uint64::ONE,
+        timestamp: Timestamp::from_nanos(100),
+        hash: Hash256::ZERO,
+    };
+
+    fn setup_test(wat: &[u8], max_gas: Option<u64>) -> (Environment, Store, Box<Instance>) {
+        let (gas_checkpoint, gas_tracker) = if let Some(max_gas) = max_gas {
+            (max_gas, GasTracker::new_limited(max_gas))
+        } else {
+            (u64::MAX, GasTracker::new_limitless())
+        };
+
+        // Compile the contract; create Wasmer store and instance.
+        let (store, instance) = {
+            let mut compiler = Singlepass::new();
+            compiler.push_middleware(Arc::new(Metering::new(0, |_| GAS_PER_OPERATION)));
+
+            let engine = Engine::from(compiler);
+            let module = Module::new(&engine, wat).unwrap();
+
+            let mut store = Store::new(engine);
+
+            let instance = Instance::new(&mut store, &module, &imports! {}).unwrap();
+            let instance = Box::new(instance);
+
+            set_remaining_points(&mut store, &instance, gas_checkpoint);
+
+            (store, instance)
+        };
+
+        // Create the function environment.
+        //
+        // For production (in `WasmVm::build_instance`) this needs to be done
+        // before creating the instance, but here for testing purpose, we don't
+        // need any import function, so this can be done later, which is simpler.
+        let env = {
+            let storage = Shared::new(MockStorage::new());
+            let storage_provider = StorageProvider::new(Box::new(storage.clone()), &[b"prefix"]);
+
+            let querier_provider = QuerierProvider::new(
+                WasmVm::new(0),
+                Box::new(storage),
+                gas_tracker.clone(),
+                MOCK_BLOCK,
+            );
+
+            let mut env = Environment::new(
+                storage_provider,
+                true,
+                querier_provider,
+                10,
+                gas_tracker,
+                gas_checkpoint,
+            );
+
+            env.set_wasmer_memory(&instance).unwrap();
+            env.set_wasmer_instance(&instance).unwrap();
+            env
+        };
+
+        (env, store, instance)
+    }
```

Comment:
I reorganized this function into sections (one for `store` and `instance`, one for `env`) for better readability.